src/qibo/_openqasm.py

```diff
@@ -0,0 +1,192 @@
+            c.add(gate)
+        self._reorder_registers(c.measurements)
+        return c
+
+    def _get_measurement(self, measurement):
+        """Converts a :class:`openqasm3.ast.QuantumMeasurementStatement` statement into a :class:`qibo.gates.measurements.M` gate."""
+        qubit = self._get_qubit(measurement.measure.qubit)
+        register = measurement.target.name.name
+        if register not in self.c_registers:
+            raise_error(ValueError, f"Undefined measurement register `{register}`.")
+        ind = measurement.target.indices[0][0].value
+        if ind >= len(self.c_registers[register]):
+            raise_error(
+                IndexError, f"Index `{ind}` is out of bounds of register `{register}`."
+            )
+        self.c_registers[register][ind] = qubit
+        return getattr(qibo.gates, "M")(qubit, register_name=register)
+
+    def _get_qubit(self, qubit):
+        """Extracts the qubit from a :class:`openqasm3.ast.QubitDeclaration` statement."""
+        if isinstance(qubit, openqasm3.ast.QubitDeclaration):
+            return qubit.qubit.name, qubit.size.value
+        if not isinstance(qubit, openqasm3.ast.Identifier):
+            return self.q_registers[qubit.name.name][qubit.indices[0][0].value]
+        else:
+            return qubit.name
+
+    def _get_gate(self, gate):
+        """Converts a :class:`openqasm3.ast.QuantumGate` statement into a :class:`qibo.gates.measurements.M` gate."""
+        qubits = [self._get_qubit(q) for q in gate.qubits]
+        init_args = []
+        for arg in gate.arguments:
+            arg = self._unroll_expression(arg)
+            try:
+                arg = eval(arg.replace("pi", "np.pi"))
+            except:
+                pass
+            init_args.append(arg)
+        try:
+            gates = [
+                getattr(qibo.gates, _qibo_gate_name(gate.name.name))(
+                    *qubits, *init_args
+                )
+            ]
+        except:
+            try:
+                gates = self.defined_gates.get(gate.name.name).get_gates(
+                    qubits, init_args
+                )
+            except:
+                raise_error(
+                    ValueError, f"Invalid gate declaration at span: {gate.span}"
+                )
+        return gates
+
+    def _unroll_expression(self, expr):
+        """Unrolls an argument definition expression to retrieve the complete argument as a string."""
+        try:
+            return expr.name
+        except:
+            try:
+                return expr.value
+            except:
+                expr_dict = {}
+                for attr in ("lhs", "op", "expression", "rhs"):
+                    expr_dict[attr] = ""
+                    try:
+                        val = self._unroll_expression(getattr(expr, attr))
+                    except:
+                        continue
+                    expr_dict[attr] += str(val)
+                return "".join(list(expr_dict.values()))
```

Comment:
Can these two functions be implemented without this many try and excepts?

Reply:
The `openqasm3` parsed objects are unfortunately a bit tricky to handle, you can get completely different objects depending on the gate declaration. For example, the two declared gates `rx(2*pi)` and `rx(pi/2)` yield to different parsed objects, which leads to different ways of reconstructing the argument. So either you hard code every different possibility, or you try the different patterns. I am open to suggestions though.